### ISSUE TITLE
Is the DiscUtil able to detect the used block and unused block of VHDX image

### ISSUE THREAD

**Issue by LunfanZhang** [DiscUtils/DiscUtils#300]
Hi @LTRData , As we know, VHDX image is dynamic allocate, basically, not all content of image is used. I am wondering is the DiscUtil able to detect and divide the content of the used block/content? and only open the used content? do we support this currently?
in the other word, do we support shrink the VHDX currently?

as blew, when we open the disk, we will get all the content stream even the most block is empty
`VirtualDisk inDisk = VirtualDisk.OpenDisk(source_file, FileAccess.Read, username, password)`
`SparseStream contentStream = inDisk.Content;`

**Comment by LTRData** [comment]
I am not sure if I follow. But if you open a dynamically allocated vhdx image, you see the entire virtual size of the image even though only parts of it are actually allocated. So, even though the image file that is opened physically allocates very little space, that is transparent to the application consuming this library, the application will see the entire virtual size.

**Comment by LunfanZhang** [comment]
Yes, I get your point @LTRData , the allocation is low-level thing. but I’m curious about that is this lib able to do such kind of low-level thing (I guess it is more like to read the extent map/meta data of VHDX) and determine which part of content/block is populated?
or is there interface to only open the allocated content? or shrink the image?

**Comment by LunfanZhang** [comment]
Hi @LTRData , let me clarify more, just like, according to: https://github.com/DiscUtils/DiscUtils/blob/59d7cadab839c6d8dfcf52f8be5efe6d2ced190f/Library/DiscUtils.Core/VirtualDisk.cs#L96
we can get the number of blocks of an VirtualDisk, but how can we identify each of block is populated or not? I did not see any related parameter or function to get it.

**Comment by LTRData** [comment]
I think I understand more about what you try to do now. If you use `Contents` property of a virtual disk object, you get a `SparseStream` object that has got an `Extents` property that you can use to enumerate allocated regions of the virtual disk.

**Comment by LunfanZhang** [comment]
Thanks, so, Can I presume it is possible to send only the Stream content between `Extent. start` to  `Extents.length` (skip the Zero content) to a certain place (e.g. raw image), it will not break the origin file system or file structure of image?

**Comment by LTRData** [comment]
> Thanks, so, Can I presume it is possible to send only the Stream content between `Extent. start` to `Extents.length` (skip the Zero content) to a certain place (e.g. raw image), it will not break the origin file system or file structure of image?

Yes, exactly!

**Comment by LunfanZhang** [comment]
Hi @LTRData , do you know except using this project DiscUtils, do we have any other way/lib to parse a VHDX/.avhdx image to get the extent information?

**Comment by LTRData** [comment]
> Hi @LTRData , do you know except using this project DiscUtils, do we have any other way/lib to parse a VHDX/.avhdx image to get the extent information? 

There is a C++ library that supports vhdx too:
https://github.com/libyal/libvhdi

I have not used it much though, but I guess it should be possible to work out extents information using that library as well.

**Comment by LunfanZhang** [comment]
Thanks for that @LTRData , and really quick response :-)
